fastedge-rust-sdk/examples/watermark/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+// this example reads file from S3 storage, which must be confgiured for the app like this:
+//   "env": {
+//    "ACCESS_KEY": "<access_key>",
+//    "BASE_HOSTNAME": "<base_hostname>, e.g. cloud.gcore.lu",
+//    "BUCKET": "<bucket>",
+//    "REGION": "<region>",
+//    "SECRET_KEY": "<secret_key>"
+//  }
+// then apply watermark from file "sample.png", which is embedded during compilation process
+// and return resulting image as PNG.
+// if file from S3 cannot be recognised as valid image, it is passed to caller as is
+
+const DEFAULT_OPACITY: f32 = 1.0;   // to use non-default opacity, specify OPACITY in 0-1.0 range in app env
+
+use std::{
+    time::Duration,
+    str::FromStr,
+    env,
+    io::{Cursor, Read, Seek, SeekFrom}
+};
+use fastedge::{
+    body::Body,
+    http::{Error, header, Request, Response, StatusCode, Method}
+};
+use mime::Mime;
+use url::Url;
+use image::*;
+use rusty_s3::{Bucket, Credentials, S3Action, UrlStyle};
+
+#[fastedge::http]
+fn main(req: Request<Body>) -> Result<Response<Body>, Error> {
+    // embed watermark file - file must be present during compilation
+    let wm_buf = include_bytes!("sample.png");
+
+    // Filter request methods
+    match req.method() {
+        // Allow only GET and HEAD requests.
+        &Method::GET | &Method::HEAD => (),
+
+        // Deny anything else.
+        _ => {
+            return Response::builder()
+                .status(StatusCode::METHOD_NOT_ALLOWED)
+                .header(header::ALLOW, "GET, HEAD")
+                .body(Body::from("This method is not allowed\n"));
+        }
+    };
+
+    // get filename from URL with has format <scheme>://<host>/<filename>
+    let filename = req
+        .uri()
+        .path()
+        .trim_start_matches("/");
+    if filename.is_empty() {
+        return Response::builder()
+            .status(StatusCode::BAD_REQUEST)
+            .body(Body::from("Malformed request - filename expected\n"));
+    }
+
+    // construct S3 signed URL
+    let (signed_url, host) = match sign_s3(filename) {
+        Err(_) => {
+            return Response::builder()
+                .status(StatusCode::INTERNAL_SERVER_ERROR)
+                .body(Body::from("App misconfigured\n"))
+        }
+        Ok((u, h)) => (u, h),
+    };
+
+    /* Actual request to S3 */
+    let s3_req = Request::builder()
+        .method(Method::GET)
+        .uri(signed_url.as_str())
+        .header("Host", host)
+        .body(Body::empty())
+        .unwrap();
```

Comment:
iI would recommend to use .expect("error message") instead for unwrap()

Reply:
changed